wright/src/grammar/parsers/expression/func_call.rs

```diff
@@ -0,0 +1,61 @@
+use crate::grammar::ast::eq::AstEq;
+use crate::grammar::ast::{Expression, FuncCallExpression};
+use crate::grammar::model::{Fragment, HasFragment};
+use crate::grammar::parsers::expression::ToExpression;
+use crate::grammar::parsers::whitespace::token_delimiter;
+use crate::grammar::parsers::with_input;
+use nom::character::complete::char as ch;
+use nom::combinator::map;
+use nom::multi::separated_list;
+use nom::sequence::{delimited, pair, terminated};
+use nom::IResult;
+
+impl<'s> FuncCallExpression<'s> {
+    /// Left parenthesis delimiting argument list. Probably should never change.
+    pub const DELIMITER_LEFT: char = '(';
+
+    /// Left parenthesis delimiting argument list. Probably should never change.
+    pub const DELIMITER_RIGHT: char = ')';
+
+    /// Comma separating arguments. Probably should never change.
+    pub const ARG_SEPARATOR: char = ',';
+
+    /// Parse an index expression in wright source code.
+    pub fn parse(input: Fragment<'s>) -> IResult<Fragment<'s>, Self> {
+        map(
+            with_input(pair(
+                terminated(
+                    Expression::parse,
+                    pair(token_delimiter, ch(Self::DELIMITER_LEFT)),
+                ),
```

Comment:
This should be its own primary parser, similar to the binary expression parser. Otherwise you would get infinite recursion.

Reply:
Oh you mean `Expression::parse` should be changed to only parse primary expressions?

Reply:
No, make a new function `FuncCall::primary` that parses the leading expression. This parser should not include binary expressions, for example.

Reply:
Why isn't the parser `BinaryExpression::primary` good enough?

```rust
    /// Parse a binary terminal symbol.
    pub fn primary(input: Fragment<'s>) -> IResult<Fragment<'s>, Expression> {
        alt((
            map(Conditional::parse, Expression::Conditional),
            map(UnaryExpression::parse, Expression::UnaryExpression),
            map(IndexExpression::parse, Expression::IndexExpression),
            map(Block::parse, Expression::Block),
            map(Parens::parse, Expression::Parens),
            map(SelfLit::parse, Expression::SelfLit),
            map(StringLit::parse, Expression::StringLit),
            map(CharLit::parse, Expression::CharLit),
            map(NumLit::parse, Expression::NumLit),
            map(BooleanLit::parse, Expression::BooleanLit),
            map(Identifier::parse, Expression::Identifier),
        ))(input)
    }
```

Is there anything here that would cause issues?

Reply:
Resolving because duplicate